File: python/src/taskulus/dependency_tree.py
"""Dependency tree rendering utilities."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Set

from taskulus.issue_files import read_issue_from_file
from taskulus.models import DependencyLink, IssueData
from taskulus.project import ProjectMarkerError, load_project_directory

MAX_TREE_NODES = 25


class DependencyTreeError(RuntimeError):
    """Raised when dependency tree rendering fails."""


@dataclass(frozen=True)
class DependencyTreeNode:
    """Dependency tree node."""

    identifier: str
    title: str
    dependency_type: Optional[str]
    dependencies: List["DependencyTreeNode"]

    def to_dict(self) -> Dict[str, object]:
        """Serialize node to a dict.

        :return: Dictionary representation.
        :rtype: Dict[str, object]
        """
        payload: Dict[str, object] = {
            "id": self.identifier,
            "title": self.title,
            "dependencies": [child.to_dict() for child in self.dependencies],
        }
        if self.dependency_type is not None:
            payload["dependency_type"] = self.dependency_type
        return payload


def build_dependency_tree(
    root: Path, identifier: str, max_depth: Optional[int]
) -> DependencyTreeNode:
    """Build a dependency tree for the given issue.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier to start from.
    :type identifier: str
    :param max_depth: Optional maximum traversal depth.
    :type max_depth: Optional[int]
    :return: Dependency tree root.
    :rtype: DependencyTreeNode
    :raises DependencyTreeError: If tree building fails.
    """
    try:
        project_dir = load_project_directory(root)
    except ProjectMarkerError as error:
        raise DependencyTreeError(str(error)) from error

    issues_dir = project_dir / "issues"
    issues = _load_issues(issues_dir)
    if identifier not in issues:
        raise DependencyTreeError("not found")

    return _build_node(
        identifier=identifier,
        issues=issues,
        max_depth=max_depth,
        depth=0,
        visited=set(),
        dependency_type=None,
    )


def render_dependency_tree(
    node: DependencyTreeNode,
    output_format: str,
    max_nodes: int = MAX_TREE_NODES,
) -> str:
    """Render a dependency tree in the requested format.

    :param node: Dependency tree root.
    :type node: DependencyTreeNode
    :param output_format: Output format (text, json, dot).
    :type output_format: str
    :param max_nodes: Maximum nodes to render for text output.
    :type max_nodes: int
    :return: Rendered output string.
    :rtype: str
    :raises DependencyTreeError: If the format is unsupported.
    """
    if output_format == "json":
        return json.dumps(node.to_dict(), indent=2)
    if output_format == "dot":
        return _render_dot(node)
    if output_format == "text":
        return _render_ascii(node, max_nodes)
    raise DependencyTreeError("invalid format")


def _load_issues(issues_dir: Path) -> Dict[str, IssueData]:
    issues: Dict[str, IssueData] = {}
    for issue_path in sorted(issues_dir.glob("*.json"), key=lambda path: path.name):
        issue = read_issue_from_file(issue_path)
        issues[issue.identifier] = issue
    return issues


def _build_node(
    identifier: str,
    issues: Dict[str, IssueData],
    max_depth: Optional[int],
    depth: int,
    visited: Set[str],
    dependency_type: Optional[str],
) -> DependencyTreeNode:
    issue = issues[identifier]
    if identifier in visited:
        return DependencyTreeNode(
            identifier=issue.identifier,
            title=issue.title,
            dependency_type=dependency_type,
            dependencies=[],
        )
    visited.add(identifier)

    dependencies: List[DependencyTreeNode] = []
    if max_depth is None or depth < max_depth:
        for dependency in issue.dependencies:
            dependencies.append(
                _build_dependency(
                    dependency=dependency,
                    issues=issues,
                    max_depth=max_depth,
                    depth=depth + 1,
                    visited=visited,
                )
            )

    return DependencyTreeNode(
        identifier=issue.identifier,
        title=issue.title,
        dependency_type=dependency_type,
        dependencies=dependencies,
    )


def _build_dependency(
    dependency: DependencyLink,
    issues: Dict[str, IssueData],
    max_depth: Optional[int],
    depth: int,
    visited: Set[str],
) -> DependencyTreeNode:
    if dependency.target not in issues:
        raise DependencyTreeError(
            f"dependency target '{dependency.target}' does not exist"
        )
    return _build_node(
        identifier=dependency.target,
        issues=issues,
        max_depth=max_depth,
        depth=depth,
        visited=visited,
        dependency_type=dependency.dependency_type,
    )


def _render_ascii(node: DependencyTreeNode, max_nodes: int) -> str:
    lines: List[str] = []
    count = 0
    truncated = False

    def visit(current: DependencyTreeNode, prefix: str, is_last: bool) -> None:
        nonlocal count, truncated
        if count >= max_nodes:
            truncated = True
            return

        if prefix:
            connector = "`-- " if is_last else "|-- "
            lines.append(f"{prefix}{connector}{current.identifier} {current.title}")
        else:
            lines.append(f"{current.identifier} {current.title}")
        count += 1

        if not current.dependencies:
            return

        child_prefix = f"{prefix}{'    ' if is_last else '|   '}"
        last_index = len(current.dependencies) - 1
        for index, child in enumerate(current.dependencies):
            visit(child, child_prefix, index == last_index)

    visit(node, "", True)

    if truncated:
        lines.append("additional nodes omitted")
    return "\n".join(lines)


def _render_dot(node: DependencyTreeNode) -> str:
    edges: List[str] = []

    def visit(current: DependencyTreeNode) -> None:
        for child in current.dependencies:
            edges.append(f'  "{current.identifier}" -> "{child.identifier}";')
            visit(child)

    visit(node)
    lines = ["digraph dependencies {", *edges, "}"]
    return "\n".join(lines)

File: rust/src/dependency_tree.rs
//! Dependency tree rendering utilities.

use std::collections::{BTreeMap, HashSet};
use std::fs;
use std::path::Path;

use serde::Serialize;

use crate::error::TaskulusError;
use crate::file_io::load_project_directory;
use crate::issue_files::read_issue_from_file;
use crate::models::{DependencyLink, IssueData};

const MAX_TREE_NODES: usize = 25;

/// Dependency tree node.
#[derive(Debug, Clone, Serialize)]
pub struct DependencyTreeNode {
    #[serde(rename = "id")]
    pub identifier: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dependency_type: Option<String>,
    pub dependencies: Vec<DependencyTreeNode>,
}

/// Build a dependency tree for the given issue.
///
/// # Arguments
/// * `root` - Repository root path.
/// * `identifier` - Issue identifier to start from.
/// * `max_depth` - Optional maximum traversal depth.
///
/// # Errors
/// Returns `TaskulusError::IssueOperation` if tree building fails.
pub fn build_dependency_tree(
    root: &Path,
    identifier: &str,
    max_depth: Option<usize>,
) -> Result<DependencyTreeNode, TaskulusError> {
    let project_dir = load_project_directory(root)?;
    let issues_dir = project_dir.join("issues");
    let issues = load_issues(&issues_dir)?;
    let issue = issues
        .get(identifier)
        .ok_or_else(|| TaskulusError::IssueOperation("not found".to_string()))?;

    build_node(issue, &issues, max_depth, 0, &mut HashSet::new(), None)
}

/// Render a dependency tree in the requested format.
///
/// # Arguments
/// * `node` - Dependency tree root.
/// * `output_format` - Output format (text, json, dot).
/// * `max_nodes` - Maximum nodes to render for text output.
///
/// # Errors
/// Returns `TaskulusError::IssueOperation` if format is unsupported.
pub fn render_dependency_tree(
    node: &DependencyTreeNode,
    output_format: &str,
    max_nodes: Option<usize>,
) -> Result<String, TaskulusError> {
    match output_format {
        "json" => {
            serde_json::to_string_pretty(node).map_err(|error| TaskulusError::Io(error.to_string()))
        }
        "dot" => Ok(render_dot(node)),
        "text" => Ok(render_ascii(node, max_nodes.unwrap_or(MAX_TREE_NODES))),
        _ => Err(TaskulusError::IssueOperation("invalid format".to_string())),
    }
}

fn load_issues(issues_dir: &Path) -> Result<BTreeMap<String, IssueData>, TaskulusError> {
    let mut issues: BTreeMap<String, IssueData> = BTreeMap::new();
    for entry in fs::read_dir(issues_dir).map_err(|error| TaskulusError::Io(error.to_string()))? {
        let entry = entry.map_err(|error| TaskulusError::Io(error.to_string()))?;
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        let issue = read_issue_from_file(&path)?;
        issues.insert(issue.identifier.clone(), issue);
    }
    Ok(issues)
}

fn build_node(
    issue: &IssueData,
    issues: &BTreeMap<String, IssueData>,
    max_depth: Option<usize>,
    depth: usize,
    visited: &mut HashSet<String>,
    dependency_type: Option<String>,
) -> Result<DependencyTreeNode, TaskulusError> {
    if visited.contains(&issue.identifier) {
        return Ok(DependencyTreeNode {
            identifier: issue.identifier.clone(),
            title: issue.title.clone(),
            dependency_type,
            dependencies: Vec::new(),
        });
    }
    visited.insert(issue.identifier.clone());

    let mut dependencies = Vec::new();
    if max_depth.map_or(true, |limit| depth < limit) {
        for dependency in &issue.dependencies {
            dependencies.push(build_dependency(
                dependency,
                issues,
                max_depth,
                depth + 1,
                visited,
            )?);
        }
    }

    Ok(DependencyTreeNode {
        identifier: issue.identifier.clone(),
        title: issue.title.clone(),
        dependency_type,
        dependencies,
    })
}

fn build_dependency(
    dependency: &DependencyLink,
    issues: &BTreeMap<String, IssueData>,
    max_depth: Option<usize>,
    depth: usize,
    visited: &mut HashSet<String>,
) -> Result<DependencyTreeNode, TaskulusError> {
    let issue = issues.get(&dependency.target).ok_or_else(|| {
        TaskulusError::IssueOperation(format!(
            "dependency target '{}' does not exist",
            dependency.target
        ))
    })?;
    build_node(
        issue,
        issues,
        max_depth,
        depth,
        visited,
        Some(dependency.dependency_type.clone()),
    )
}

fn render_ascii(node: &DependencyTreeNode, max_nodes: usize) -> String {
    let mut lines: Vec<String> = Vec::new();
    let mut count = 0;
    let mut truncated = false;

    fn visit(
        current: &DependencyTreeNode,
        prefix: &str,
        is_last: bool,
        lines: &mut Vec<String>,
        count: &mut usize,
        truncated: &mut bool,
        max_nodes: usize,
    ) {
        if *count >= max_nodes {
            *truncated = true;
            return;
        }

        if prefix.is_empty() {
            lines.push(format!("{} {}", current.identifier, current.title));
        } else {
            let connector = if is_last { "`-- " } else { "|-- " };
            lines.push(format!(
                "{}{}{} {}",
                prefix, connector, current.identifier, current.title
            ));
        }
        *count += 1;

        if current.dependencies.is_empty() {
            return;
        }

        let child_prefix = format!("{}{}", prefix, if is_last { "    " } else { "|   " });
        let last_index = current.dependencies.len().saturating_sub(1);
        for (index, child) in current.dependencies.iter().enumerate() {
            visit(
                child,
                &child_prefix,
                index == last_index,
                lines,
                count,
                truncated,
                max_nodes,
            );
        }
    }

    visit(
        node,
        "",
        true,
        &mut lines,
        &mut count,
        &mut truncated,
        max_nodes,
    );
    if truncated {
        lines.push("additional nodes omitted".to_string());
    }
    lines.join("\n")
}

fn render_dot(node: &DependencyTreeNode) -> String {
    let mut edges: Vec<String> = Vec::new();

    fn visit(current: &DependencyTreeNode, edges: &mut Vec<String>) {
        for child in &current.dependencies {
            edges.push(format!(
                "  \"{}\" -> \"{}\";",
                current.identifier, child.identifier
            ));
            visit(child, edges);
        }
    }

    visit(node, &mut edges);
    let mut lines = Vec::new();
    lines.push("digraph dependencies {".to_string());
    lines.extend(edges);
    lines.push("}".to_string());
    lines.join("\n")
}
